### PR TITLE
Add sparse_vectorized option for LindbladModel

### PR DESCRIPTION
**Pull request by averyparr**
### Summary

Adds new evaluation_mode 'sparse_vectorized' to LindbladModel, supported by a new SparseLindbladCollection.

### CODE REVIEW

**Review by DanPuzzuoli — CHANGES_REQUESTED** [review]
Looks good to me, just:
- Delete the `print` line.
- There is some spurious empty file `qiskit_dynamics/models/tmp.ipynb` that is appearing in the list of changes. Can you delete this.

Are there any existing tests for the behaviour of `to_array`? (This may have been done in your last PR, just wondering.)

File: qiskit_dynamics/models/rotating_frame.py [hunk 1]
```diff
@@ -506,32 +506,37 @@ def vectorized_map_into_frame(
         if self.frame_diag is not None:
             # Put the vectorized operator into the frame basis
             if not operator_in_frame_basis and self.frame_basis is not None:
+                print("Should Not 1")
```
[the remark this answers is not in the document]
Comment (averyparr) [reply qiskit_dynamics/models/rotating_frame.py]: There were no (direct) existing checks for to_array of which I am aware–there were plenty of indirect tests. In any case, I've added a test in a more recent commit.

**Review by DanPuzzuoli — COMMENTED** [review]
I think one outstanding issue currently is that we can't define models using sparse matrices. E.g. I just tried doing: `HamiltonianModel(operators=[sparse1, sparse2])` (for `sparsex` being `scipy` `csr_matrix` instances) and an error gets raised by `to_array` as it doesn't properly handle a list of sparse ops.

I think probably this connects with a bunch of things:
- Hamiltonian validation (does this work for sparse?)
- Instantiation of operator collections - I think if operators are passed as sparse then when `to_array` is called in `DenseOperatorCollection`, under the current behaviour we've defined where it "leaves sparse things alone too" it'll won't output the correct shape.

I don't think this _necessarily_ needs to be solved by this PR as it's somewhat unexpected and isn't exactly what this PR is "about", but we should think about what a solution looks like and then, at the very least, modify this PR with an intermediate fix that avoids these errors.

For a solution:
- I think the problem design-wise is that `to_array` isn't really doing what the name suggests anymore (which is due to my own naive initial suggestion to get it to handle matrices as well). It's making too many decisions on output type based on input type.
- We probably need to revert `to_array` back to something that _only_ outputs `Array`s, and then make a `to_csr` that outputs either a CSR or list of CSRs. These can be used by the respective `OperatorCollection`s and there won't be any ambiguity on output type based on input type, which is really what the `OperatorCollection`s need: they need to know exactly what they're getting.
- The above still leaves the problem of validation open: I believe the Hamiltonian validation code assumes that it's getting a 3d array - but if we want the user to be able to directly pass in sparse matrices, and never have them converted to dense, we need to make sure this validation step works _without_ the initial call to `to_array`. 

I think one potential temporary solution for this PR that doesn't stand in the way of a "proper" solution is:
- Just revert `to_array` back to always outputting an `Array` (even if CSR marices are passed). This will break the desired "sparse stays sparse" functionality (if a user specifies things in terms of sparse arrays and wants the evaluation mode to be sparse), but this doesn't currently work anyway.
- I can then make a new issue describing these type handling concerns that can be handled in a separate PR.

Thoughts?

**Review by DanPuzzuoli — APPROVED** [review]
This looks good to me, thanks!

**Pull request merged**